src/lib.rs

```diff
@@ -50,25 +50,23 @@ impl Preprocessor for D2 {
                 cmark(events, &mut buf).unwrap();
                 chapter.content = buf;
             }
-        });
-
-        Ok(book)
+        })
     }
 }
 
 fn process_events<'a>(
     backend: &'a Backend,
     chapter: &'a Chapter,
     events: impl Iterator<Item = Event<'a>> + 'a,
-) -> impl Iterator<Item = Event<'a>> + 'a {
+) -> Result<impl Iterator<Item = Event<'a>> + 'a, Error> {
     let mut in_block = false;
     // if Windows crlf line endings are used, a code block will consist
     // of many different Text blocks, thus we need to buffer them in here
     // see https://github.com/raphlinus/pulldown-cmark/issues/507
     let mut diagram = String::new();
     let mut diagram_index = 0;
 
-    events.flat_map(move |event| {
+    Ok(events.try_for_each(move |event| {
```

Comment:
try_for_each isn't what you're looking for here. `try_for_each` will run the closure on each element in the iterator, and then return nothing. What you want is to do is process each element in the iterator, and then pass it along. You'll want to use the 'map' family of iterator adaptors.

I've used `flat_map` in the original implementation. It takes an iterator, runs each element through a closure that returns an iterator, and then 'flattens' the inner iterators by chaining them all together

Have a read through the iterator docs - https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html

Reply:
I reverted to flat_map because none of the other iterators listed seemed to allow me to return a Result